Add tx metadata to sqlite database

rvl opened the pull request:
### Issue Number

ADP-307 / #2072 

### Overview

Includes transaction metadata in the sqlite database transaction history.

### Comments

- Based on PR #2079 branch
- DB state machine tests fail. Possibly due to mutations during the TxMetadata json round-trip.

rvl committed:
Add tx metadata to sqlite database

rvl committed:
Shrinker for TxMetadata

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -128,13 +129,13 @@ import Data.Either
     ( (^.) )
 import Data.List
-    ( nub, sortOn, unzip3 )
+    ( nub, sortOn, unzip4 )
```
I like how it started with just unzip2 and gradually increased. We still have 3 left, after which we have to stop further development :)

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1252,30 +1256,34 @@ updateTxMetas
         [ (TxOutKey txOutputTxId txOutputIndex, o)
         | o@TxOut{..} <- txouts ]
+    forM_ (chunksOf chunkSize $ map txMetaTxId metas) $ \txids ->
+        deleteWhere [ TxMetadataTxId <-. txids ]
```
Why the delete? Isn't the point of `repsert` to "replace or insert"?

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1039,13 +1039,14 @@ mkTxHistory wid txs = flatTxHistory
+        , concatMap (snd . (\(_,b,_,_) -> b)) entities
+        , concatMap (\(_,_,c,_) -> maybeToList c) entities
+        , concatMap (\(_,_,_,d) -> d) entities
         )
```
Wouldn't be more efficient and simpler to store metadata with the rest of the _static_ data (status, direction etc..)? 

We have to resort to separate database tables for TxIn, TxOut and TxWithdrawals because they are unbounded (modulo tx size) **lists**. Metadata however can be treated as a bunch of bytes that are either present or not.

rvl replied:
My reason was just that sqlite tables don't cost much and it seemed like it would be confusing to have data in the `TxMeta` table which was actually a field of `Tx`.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -100,6 +100,16 @@ TxMeta
+    txMetadataValue   W.TxMetadata  sql=value
+
+    Primary txMetadataTxId
+    deriving Show Generic
```
I am a bit doubtful about having this as a separate table (cf earlier comment).

KtorZ committed:
store transaction metadata alongside other derived informations

  This is more efficient than storing them in separate tables like inputs and outputs and also a bit less awkward when it comes to fetching metadata corresponding to each transactions back.

KtorZ committed:
temporarily work-around jsonToMetadata limitations in state-machine tests

KtorZ reviewed (approved):
Temporarily worked-around the roundtripping issues until #2098 is fixed. state-machine tests pass for now.

rvl committed:
hlint
